File: cmd/watch/web/src/lib/protocol/viewerProtocol.ts
/**
 * Type definitions and normalization functions for the SSE graph stream protocol.
 * These functions validate and normalize untrusted JSON payloads from the server.
 */

export interface WorktreeDescriptor {
  id: string;
  path: string;
  label: string;
  isPrimary: boolean;
  // False once the underlying git worktree is removed — the tab becomes a
  // frozen, closable record. A missing flag is treated as active.
  active: boolean;
}

export interface Snapshot {
  id: number;
  worktreeId?: string;
  timestamp: string;
  dot: string;
  // True for the first snapshot recorded for a worktree this session — the
  // state present when the watcher attached. Omitted (falsy) otherwise.
  sessionStart?: boolean;
}

export interface CommitSummary {
  hash: string;
  shortHash: string;
  subject: string;
  author: string;
  email: string;
  timestamp: string;
}

export interface Collection {
  id: number;
  worktreeId?: string;
  timestamp: string;
  snapshots: Snapshot[];
  commitHistory: CommitSummary[];
}

export interface GraphStreamPayload {
  worktrees?: WorktreeDescriptor[];
  // Session-global render format of every snapshot's `dot` field ("dot" or
  // "mermaid"). Absent on older payloads; treated as "dot".
  format?: string;
  workingSnapshots: Snapshot[];
  pastCollections: Collection[];
  latestWorkingId?: number;
  latestPastCollectionId?: number;
}

function normalizeWorktree(worktree: unknown): WorktreeDescriptor | null {
  if (!worktree || typeof worktree !== "object") {
    return null;
  }
  const r = worktree as Record<string, unknown>;
  if (typeof r.id !== "string" || r.id === "") {
    return null;
  }
  return {
    id: r.id,
    path: typeof r.path === "string" ? r.path : "",
    label: typeof r.label === "string" ? r.label : r.id,
    isPrimary: r.isPrimary === true,
    // Treat a missing flag as active so older payloads keep their tabs pinned.
    active: r.active !== false,
  };
}

function normalizeSnapshot(snapshot: unknown): Snapshot | null {
  if (!snapshot || typeof snapshot !== "object") {
    return null;
  }
  const s = snapshot as Record<string, unknown>;
  if (typeof s.dot !== "string") {
    return null;
  }

  const normalized: Snapshot = {
    id: Number.isFinite(s.id) ? (s.id as number) : 0,
    worktreeId: typeof s.worktreeId === "string" ? s.worktreeId : "",
    timestamp: typeof s.timestamp === "string" ? s.timestamp : new Date(0).toISOString(),
    dot: s.dot,
  };
  // Preserve the marker only when set, mirroring the backend's omitempty so
  // unmarked snapshots stay free of the field.
  if (s.sessionStart === true) {
    normalized.sessionStart = true;
  }
  return normalized;
}

function normalizeCommitSummary(commit: unknown): CommitSummary | null {
  if (!commit || typeof commit !== "object") {
    return null;
  }
  const c = commit as Record<string, unknown>;
  if (typeof c.hash !== "string" || c.hash === "") {
    return null;
  }
  return {
    hash: c.hash,
    shortHash: typeof c.shortHash === "string" ? c.shortHash : c.hash.slice(0, 7),
    subject: typeof c.subject === "string" ? c.subject : "",
    author: typeof c.author === "string" ? c.author : "",
    email: typeof c.email === "string" ? c.email : "",
    timestamp: typeof c.timestamp === "string" ? c.timestamp : new Date(0).toISOString(),
  };
}

function normalizeCollection(collection: unknown): Collection | null {
  if (!collection || typeof collection !== "object") {
    return null;
  }
  const c = collection as Record<string, unknown>;
  if (!Array.isArray(c.snapshots)) {
    return null;
  }

  return {
    id: Number.isFinite(c.id) ? (c.id as number) : 0,
    worktreeId: typeof c.worktreeId === "string" ? c.worktreeId : "",
    timestamp: typeof c.timestamp === "string" ? c.timestamp : new Date(0).toISOString(),
    snapshots: c.snapshots
      .map(normalizeSnapshot)
      .filter((snapshot): snapshot is Snapshot => snapshot !== null),
    commitHistory: Array.isArray(c.commitHistory)
      ? c.commitHistory.map(normalizeCommitSummary).filter((commit): commit is CommitSummary => commit !== null)
      : [],
  };
}

/**
 * Normalizes untrusted SSE JSON payloads from the watch server.
 * Keeps only fields needed by the viewer state machine.
 */
export function normalizeGraphStreamPayload(payload: unknown): GraphStreamPayload {
  if (!payload || typeof payload !== "object") {
    return {
      worktrees: [],
      format: "dot",
      workingSnapshots: [],
      pastCollections: [],
    };
  }

  const p = payload as Record<string, unknown>;

  return {
    worktrees: Array.isArray(p.worktrees)
      ? p.worktrees.map(normalizeWorktree).filter((worktree): worktree is WorktreeDescriptor => worktree !== null)
      : [],
    workingSnapshots: Array.isArray(p.workingSnapshots)
      ? p.workingSnapshots.map(normalizeSnapshot).filter((snapshot): snapshot is Snapshot => snapshot !== null)
      : [],
    pastCollections: Array.isArray(p.pastCollections)
      ? p.pastCollections.map(normalizeCollection).filter((collection): collection is Collection => collection !== null)
      : [],
    latestWorkingId: Number.isFinite(p.latestWorkingId) ? (p.latestWorkingId as number) : 0,
    latestPastCollectionId: Number.isFinite(p.latestPastCollectionId) ? (p.latestPastCollectionId as number) : 0,
    format: typeof p.format === "string" && p.format !== "" ? p.format : "dot",
  };
}

File: cmd/watch/protocol/viewer_protocol.go
package protocol

import "time"

const (
	RouteIndex  = "/"
	RouteEvents = "/events"
	// RouteCloseWorktree tears down a finished (inactive) worktree tab. The
	// {id} path value is the WorktreeDescriptor ID. This is the one
	// client→server message in the watch protocol; everything else flows
	// server→client.
	RouteCloseWorktree = "POST /worktrees/{id}/close"
)

const SSEEventGraph = "graph"

// WorktreeKind classifies a worktree using git's own vocabulary: the single
// main worktree created by `git init`/`git clone`, or one of possibly several
// linked worktrees created via `git worktree add`. See git-worktree(1). Kept
// independent of vcs/git.WorktreeKind so this package stays a self-contained
// wire contract.
type WorktreeKind string

const (
	WorktreeKindMain   WorktreeKind = "main"
	WorktreeKindLinked WorktreeKind = "linked"
)

// WorktreeDescriptor names a working tree visible to the watch session.
// One descriptor per tab in the viewer.
type WorktreeDescriptor struct {
	// ID is a stable identifier for the tree across reconnects.
	// "main" for the repository's main worktree; "wt-<hash8>" otherwise.
	ID string `json:"id"`
	// Path is the absolute path to the working tree on disk.
	Path string `json:"path"`
	// Label is a human-readable name for tab display.
	Label string `json:"label"`
	// Kind is WorktreeKindMain for the repository's main worktree,
	// WorktreeKindLinked otherwise.
	Kind WorktreeKind `json:"kind"`
	// Active reports whether the worktree is still being watched. It flips to
	// false when the underlying git worktree is removed: the tab stays visible
	// as a frozen, read-only record and becomes user-closable. The main
	// worktree stays active for the life of the watch session.
	Active bool `json:"active"`
}

// GraphSnapshot is the atom in the watch protocol timeline.
type GraphSnapshot struct {
	ID         int64     `json:"id"`
	WorktreeID string    `json:"worktreeId"`
	Timestamp  time.Time `json:"timestamp"`
	DOT        string    `json:"dot"`
	// SessionStart marks the first snapshot recorded for a worktree in this
	// watch session — the state that already existed when the watcher attached.
	// Changes made before this point are not in the timeline. Set once per repo
	// and preserved across commit/archive cycles.
	SessionStart bool `json:"sessionStart,omitempty"`
}

// CommitSummary describes a commit that caused a working snapshot cycle to
// freeze into an archived collection.
type CommitSummary struct {
	Hash      string    `json:"hash"`
	ShortHash string    `json:"shortHash"`
	Subject   string    `json:"subject"`
	Author    string    `json:"author"`
	Email     string    `json:"email"`
	Timestamp time.Time `json:"timestamp"`
}

// GraphStreamPayload is the wire payload for SSE "graph" events.
type GraphStreamPayload struct {
	Worktrees []WorktreeDescriptor `json:"worktrees"`
	// Format is the render format of every snapshot's DOT field for this watch
	// session ("dot" or "mermaid"). Session-global; clients treat an empty value
	// as "dot".
	Format                 string               `json:"format,omitempty"`
	WorkingSnapshots       []GraphSnapshot      `json:"workingSnapshots"`
	PastCollections        []SnapshotCollection `json:"pastCollections"`
	LatestWorkingID        int64                `json:"latestWorkingId"`
	LatestPastCollectionID int64                `json:"latestPastCollectionId"`
}

// SnapshotCollection represents an archived batch of working snapshots.
type SnapshotCollection struct {
	ID            int64           `json:"id"`
	WorktreeID    string          `json:"worktreeId"`
	Timestamp     time.Time       `json:"timestamp"`
	Snapshots     []GraphSnapshot `json:"snapshots"`
	CommitHistory []CommitSummary `json:"commitHistory,omitempty"`
}
